src/parse.cpp

```diff
@@ -287,6 +287,15 @@ expression* parser::parse_primary_expression(precedence prec) {
             function_attributes::normal, left_paren_span.begin(),
             /*allow_in_operator=*/prec.in_operator);
         return ast;
+      } else if ((this->peek().type == token_type::left_curly) &&
+                 (!this->peek().has_leading_newline)) {
+        this->error_reporter_->report(
+            error_missing_arrow_operator_in_arrow_function{
+                .where = this->peek().span()});
+        expression* ast = this->parse_arrow_function_body(
+            function_attributes::normal, this->peek().span().begin(),
+            /*allow_in_operator=*/prec.in_operator);
```

Comment:
Nit: Duplicating these lines feels weird to me. What about something like this to avoid the duplication:
```c++
      bool is_arrow_function = this->peek().type == token_type::equal_greater;
      bool is_arrow_function_without_arrow =
          (this->peek().type == token_type::left_curly) &&
          (!this->peek().has_leading_newline);
      if (is_arrow_function || is_arrow_function_without_arrow) {
        // Arrow function: () => expression-or-block
        // Arrow function: () { }  // Invalid.
        if (is_arrow_function) {
          this->skip();
        } else {
          this->error_reporter_->report(
              error_missing_arrow_operator_in_arrow_function{
              .where = this->peek().span()});
        }
        expression* ast = this->parse_arrow_function_body(
            function_attributes::normal, left_paren_span.begin(),
            /*allow_in_operator=*/prec.in_operator);
        return ast;
      } else {
        // ()  // Invalid.
        this->error_reporter_->report(
            error_missing_expression_between_parentheses{
                .left_paren_to_right_paren = source_code_span(
                    left_paren_span.begin(), right_paren_span.end()),
                .left_paren = left_paren_span,
                .right_paren = right_paren_span,
            });
        expression* child = this->parse_expression();
        return child;
      }
```